src/scanner.c

```diff
@@ -481,6 +481,13 @@ ts_scanner_scan(ScannerCtx *ctx)
 				ts_scanner_rescan(ctx, NULL);
 				continue;
 			}
+			else if (scan_result == SCAN_RESTART_WITH_NEW_SNAPSHOT)
+			{
+				ts_scanner_end_scan(ctx);
+				ctx->internal.tinfo.count = 0;
+				ctx->snapshot = GetLatestSnapshot();
+				ts_scanner_start_scan(ctx);
+			}
```

Comment:
Do you think this can be done by the caller? When I have to debug something related to the scanner, I often find it easier to just rewrite the code from scratch using the PG scan functions, because I'm unable to follow locking or snapshot behavior through all these callbacks. Now if it's going to even re-take the snapshot under some circumstances, it's going to become even more complicated.

Reply:
@jnidzwetzki  Introducing this new `SCAN_RESTART_WITH_NEW_SNAPSHOT` you're making `SCAN_RESCAN` useless, I mean no other place is using it

Reply:
@fabriziomello The `SCAN_RESCAN` option is now removed.

Reply:
@akuzm In general, this should be possible. However, we would need to introduce some logic into the scanner to bubble up the information that we have hit an updated tuple during the scan. 

Personally, I like to encapsulate this kind of generic logic into the scanner since we might need this more often in the future if we use row-locks more frequently. Do you have a general concern about this apart from adding more logic to the complicated scanner code?

Reply:
> Do you have a general concern about this apart from adding more logic to the complicated scanner code?

No, I think it's OK, at least more correct than what we have now. Generally the `GetLatestSnapshot()` can have tricky interaction with `repeatable read` or `serializable` transaction isolation levels, because it kind of downgrades to `read committed`. But I don't know this part of code well enough.